src/box/sql/func.c

```diff
@@ -2336,3 +2341,133 @@ static struct func_vtab func_sql_builtin_vtab = {
 	.call = func_sql_builtin_call_stub,
 	.destroy = func_sql_builtin_destroy,
 };
+
+/** Table of methods of SQL user-defined functions. */
+static struct func_vtab func_sql_expr_vtab;
+
+/** SQL user-defined function. */
+struct func_sql_expr {
+	/** Function object base class. */
+	struct func base;
+	/** Prepared SQL statement. */
+	struct Vdbe *stmt;
+};
+
+struct func *
+func_sql_expr_new(struct func_def *def)
+{
+	struct sql *db = sql_get();
+	const char *body = def->body;
+	uint32_t body_len = body == NULL ? 0 : strlen(body);
+	struct Expr *expr = sql_expr_compile(db, body, body_len);
+	if (expr == NULL)
+		return NULL;
+
+	struct Parse parser;
+	sql_parser_create(&parser, db, default_flags);
+	struct Vdbe *v = sqlGetVdbe(&parser);
+	if (v == NULL) {
+		sql_parser_destroy(&parser);
+		sql_expr_delete(db, expr);
+		return NULL;
+	}
+	int ref_reg = ++parser.nMem;
+	sqlVdbeAddOp2(v, OP_Variable, ++parser.nVar, ref_reg);
+	parser.vdbe_field_ref_reg = ref_reg;
+
+	sqlVdbeSetNumCols(v, 1);
+	vdbe_metadata_set_col_name(v, 0, def->name);
+	vdbe_metadata_set_col_type(v, 0, field_type_strs[def->returns]);
+	int res_reg = sqlExprCodeTarget(&parser, expr, ++parser.nMem);
+	sqlVdbeAddOp2(v, OP_ResultRow, res_reg, 1);
+
+	bool is_error = parser.is_aborted;
+	sql_finish_coding(&parser);
+	sql_parser_destroy(&parser);
+	sql_expr_delete(db, expr);
+
+	if (is_error) {
+		sql_stmt_finalize((struct sql_stmt *)v);
+		return NULL;
+	}
+	struct func_sql_expr *func = xmalloc(sizeof(*func));
+	func->stmt = v;
+	func->base.vtab = &func_sql_expr_vtab;
+	func->base.def = def;
```

Comment:
You don't need to assign `def` - it is done in `func_new()` already.

Reply:
Fixed.